index.bs

```diff
@@ -2302,6 +2344,45 @@ To <dfn>parse trigger specs</dfn> given a [=map=] |map|, a [=moment=]
 Issue: Invoke [=parse summary buckets=] and [=parse summary window operator=]
 from this algorithm.
 
+To <dfn>parse max event states</dfn> from a [=map=] |map|,
+a possibly null 32-bit positive integer |attributionScopeLimit|:
+1. If |attributionScopeLimit| is null:
+    1. If |map|["<code>[=source-registration JSON key/max_event_states=]</code>"] [=map/exists=], return an error.
+    1. Return null.
+1. Let |maxEventStates| be the result of running
+    [=parse an optional 64-bit signed integer=] with |map|,
+    "<code>[=source-registration JSON key/max_event_states=]</code>", and [=default max event states=].
+1. If |maxEventStates| is greater than the user agent's [=max trigger-state cardinality=], return an error.
+1. If |maxEventStates| is less than or equal to 0, return an error.
+1. Return |maxEventStates|.
+
+To <dfn>parse attribution scopes for source</dfn> from a [=map=] |map| and a possibly null 32-bit positive integer |attributionScopeLimit|:
+1. Let |result| be a new [=set=].
+1. If |map|["<code>[=source-registration JSON key/attribution_scopes=]</code>"] does not [=map/exist=], return |result|.
+1. Let |values| be |map|["<code>[=source-registration JSON key/attribution_scopes=]</code>"].
+1. If |values| is not a [=list=], return an error.
+1. [=list/iterate|For each=] |value| of |values|:
+    1. If |value| is not a [=string=], return an error.
+    1. If |value|'s [=string/length=] is greater than [=max length per attribution scope=], return an error.
+    1. [=set/Append=] |value| to |result|.
+1. If |result| is not [=set/is empty|empty=]:
+    1. If |attributionScopeLimit| is null, return an error.
+1. If |result|'s [=set/size=] is greater than |attributionScopeLimit|, return an error.
+1. If |result|'s [=set/size=] is greater than [=max attribution scopes per source=], return an error.
+1. Return |result|.
+
+To <dfn>parse attribution scopes for trigger</dfn> from a [=map=] |map|:
+1. Let |result| be a new [=set=].
+1. If |map|["<code>[=trigger-registration JSON key/attribution_scopes=]</code>"] does not [=map/exist=], return null.
+1. Let |values| be |map|["<code>[=source-registration JSON key/attribution_scopes=]</code>"].
+1. If |values| is not a [=list=], return an error.
+1. [=list/iterate|For each=] |value| of |values|:
+    1. If |value| is not a [=string=], return an error.
+    1. If |value|'s [=string/length=] is greater than [=max length per attribution scope=], return an error.
+    1. [=set/Append=] |value| to |result|.
+1. If |result| is [=set/is empty|empty=], return null.
```

Comment:
Is this supposed to be null or an error?

Reply:
Per your comment in the explainer, we can treat empty list the same as null or omitted.

Reply:
I'm confused now, as the header validator will return an empty set if the JSON list is empty, but the spec returns null if the JSON list is empty.

Can we make the spec and header validator agree?

Reply:
Done. now we will check trigger attribution scope emptiness at attribution time in the spec.